### PR TITLE
fix: deleted api token can be reused if created again with same name

### LINKED ISSUE #4960: Bug: [Prod] In a specific case we are able to use old deleted API token
### 📜 Description

when an api token is deleted, and then created again with the same name, we are able to use the old token (which was deleted) to access devtron

### 👟 Reproduction steps

Create a api token with name superadmin ->copy the token -> if someone creates token with same name again we are able to use hit api with old superadmin token value

### 👍 Expected behavior

old (deleted) token should give "invalid user"

### 👎 Actual Behavior

successfully authenticates via deleted token

### ☸ Kubernetes version

.

### Cloud provider

<details>
.
</details>

### 🌍 Browser

Chrome

### 🧱 Your Environment

_No response_

### ✅ Proposed Solution

_No response_

### 👀 Have you spent some time to check if this issue has been raised before?

- [X] I checked and didn't find any similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/devtron-labs/devtron/blob/main/CODE_OF_CONDUCT.md)

AB#9416

### PR DESCRIPTION
**Pull request by ashishdevtron**
# Description

introduced api token versioning to keep track of deleted/old tokens

Fixes #4960

## Checklist:

* [ ] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] Does this PR requires documentation updates?
* [ ] I've updated documentation as required by this PR.
* [ ] I have performed a self-review of my own code.
* [ ] I have commented my code, particularly in hard-to-understand areas.
* [ ] I have tested it for all user roles.
* [ ] I have added all the required unit/api test cases.

## Does this PR introduce a user-facing change?

```release-note

```

### CODE REVIEW

File: pkg/apiToken/ApiTokenRepository.go [hunk 1]
```diff
@@ -60,6 +63,20 @@ func (impl ApiTokenRepositoryImpl) Update(apiToken *ApiToken) error {
 	return impl.dbConnection.Update(apiToken)
 }
 
+func (impl ApiTokenRepositoryImpl) UpdateIf(apiToken *ApiToken, previousVersion int) error {
+	res, err := impl.dbConnection.Model(apiToken).
+		Where("name = ?", apiToken.Name).
+		Where("version = ?", previousVersion).
```
Comment (nishant-d) [inline pkg/apiToken/ApiTokenRepository.go:69]: use ID in where clause

File: pkg/apiToken/ApiTokenRepository.go [hunk 2]
```diff
@@ -60,6 +63,20 @@ func (impl ApiTokenRepositoryImpl) Update(apiToken *ApiToken) error {
...
+		Update()
+	if err != nil {
+		return err
+	}
+	if res.RowsAffected() == 0 {
+		return fmt.Errorf(ConcurrentTokenUpdateRequest)
```
Comment (nishant-d) [inline pkg/apiToken/ApiTokenRepository.go:75]: version not found/ version mismatch

File: pkg/apiToken/ApiTokenService.go [hunk 3]
```diff
@@ -181,11 +180,21 @@ func (impl ApiTokenServiceImpl) CreateApiToken(request *openapi.CreateApiTokenRe
 
 	impl.logger.Info(fmt.Sprintf("apiTokenExists : %s", strconv.FormatBool(apiTokenExists)))
 
-	// step-2 - Build email
-	email := fmt.Sprintf("%s%s", API_TOKEN_USER_EMAIL_PREFIX, name)
+	// step-2 - Build email and version
+	email := fmt.Sprintf("%s%s", userBean.API_TOKEN_USER_EMAIL_PREFIX, name)
+	var (
+		tokenVersion    int
+		previousVersion int
```
Comment (nishant-d) [inline pkg/apiToken/ApiTokenService.go:187]: previousTokenVersion

File: pkg/apiToken/ApiTokenService.go [hunk 4]
```diff
@@ -214,21 +223,30 @@ func (impl ApiTokenServiceImpl) CreateApiToken(request *openapi.CreateApiTokenRe
...
+		Version:      tokenVersion,
 		AuditLog:     sql.AuditLog{UpdatedOn: time.Now()},
 	}
 	if apiTokenExists {
 		apiTokenSaveRequest.Id = apiToken.Id
 		apiTokenSaveRequest.CreatedBy = apiToken.CreatedBy
 		apiTokenSaveRequest.CreatedOn = apiToken.CreatedOn
 		apiTokenSaveRequest.UpdatedBy = createdBy
-		err = impl.apiTokenRepository.Update(apiTokenSaveRequest)
+		err = impl.apiTokenRepository.UpdateIf(apiTokenSaveRequest, previousVersion)
```
Comment (nishant-d) [inline pkg/apiToken/ApiTokenService.go:234]: add comment

File: pkg/auth/user/UserAuthService.go [hunk 5]
```diff
@@ -488,22 +490,37 @@ func (impl UserAuthServiceImpl) AuthVerification(r *http.Request) (bool, error)
 		}
 		return false, err
 	}
+	if strings.HasPrefix(emailId, userBean.API_TOKEN_USER_EMAIL_PREFIX) {
+		tokenName := helper.ExtractTokenNameFromEmail(emailId)
+		embeddedTokenVersion, _ := strconv.Atoi(version)
+		isProvidedTokenValid, err := impl.userRepository.CheckIfUserIsValidByTokenNameAndVersion(tokenName, embeddedTokenVersion)
```
Comment (nishant-d) [inline pkg/auth/user/UserAuthService.go:496]: backward compatibility

File: pkg/auth/user/UserService.go [hunk 6]
```diff
@@ -1237,6 +1239,20 @@ func (impl *UserServiceImpl) GetUserByToken(context context.Context, token strin
 		}
 		return http.StatusUnauthorized, "", err
 	}
+	if userInfo.UserType == bean.USER_TYPE_API_TOKEN {
```
Comment (nishant-d) [inline pkg/auth/user/UserService.go:1242]: make it common function

**Review by nishant-d — CHANGES_REQUESTED** [review]

File: pkg/apiToken/ApiTokenRepository.go (hunk 1 above)
Comment (ashishdevtron) [reply pkg/apiToken/ApiTokenRepository.go]: done

File: pkg/apiToken/ApiTokenRepository.go (hunk 2 above)
Comment (ashishdevtron) [reply pkg/apiToken/ApiTokenRepository.go]: done

File: pkg/apiToken/ApiTokenService.go (hunk 3 above)
Comment (ashishdevtron) [reply pkg/apiToken/ApiTokenService.go]: done

File: pkg/apiToken/ApiTokenService.go (hunk 4 above)
Comment (ashishdevtron) [reply pkg/apiToken/ApiTokenService.go]: done

File: pkg/auth/user/UserAuthService.go (hunk 5 above)
Comment (ashishdevtron) [reply pkg/auth/user/UserAuthService.go]: done

File: pkg/auth/user/UserService.go (hunk 6 above)
Comment (ashishdevtron) [reply pkg/auth/user/UserService.go]: done

File: pkg/auth/user/UserService.go [hunk 7]
```diff
@@ -1237,6 +1238,18 @@ func (impl *UserServiceImpl) GetUserByToken(context context.Context, token strin
 		}
 		return http.StatusUnauthorized, "", err
 	}
+	if userInfo.UserType == bean.USER_TYPE_API_TOKEN && len(version) > 0 {
```
Comment (nishant-d) [inline pkg/auth/user/UserService.go:1241]: add comment around length

File: pkg/auth/user/repository/UserRepository.go [hunk 8]
```diff
@@ -242,3 +243,14 @@ func (impl UserRepositoryImpl) GetCountExecutingQuery(query string) (int, error)
 	}
 	return totalCount, err
 }
+
+func (impl UserRepositoryImpl) CheckIfTokenIsValidByTokenNameAndVersion(tokenName string, tokenVersion int) (bool, error) {
+	dbTokenVersion := 0
+	query := "SELECT t.version FROM api_token t WHERE t.name = ?"
```
Comment (nishant-d) [inline pkg/auth/user/repository/UserRepository.go:249]: why native query ?

File: pkg/auth/user/repository/UserRepository.go [hunk 9]
```diff
@@ -242,3 +243,14 @@ func (impl UserRepositoryImpl) GetCountExecutingQuery(query string) (int, error)
...
+	_, err := impl.dbConnection.Query(&dbTokenVersion, query, tokenName)
+	if err != nil {
+		impl.Logger.Error(err)
+		return false, err
+	}
+	return dbTokenVersion == tokenVersion, nil
```
Comment (nishant-d) [inline pkg/auth/user/repository/UserRepository.go:255]: we can add version comparison inside where clause itself

**Review by nishant-d — CHANGES_REQUESTED** [review]

File: pkg/auth/user/UserService.go [hunk 10]
```diff
@@ -1237,6 +1238,21 @@ func (impl *UserServiceImpl) GetUserByToken(context context.Context, token strin
 		}
 		return http.StatusUnauthorized, "", err
 	}
+	// checking length of version, to ensure backward compatibility as earlier we did not
+	// have version for api-tokens
+	// therefore, for tokens without version we will skip the below part
+	if userInfo.UserType == bean.USER_TYPE_API_TOKEN && len(version) > 0 {
+		isProvidedTokenValid, err := userHelper.CheckIfTokenIsValid(email, version, impl.userRepository)
+		if err != nil || !isProvidedTokenValid {
```
Comment (nishant-d) [inline pkg/auth/user/UserService.go:1246]: duplicate code

File: pkg/auth/user/helper/helper.go [hunk 11]
```diff
@@ -44,3 +47,14 @@ func CheckIfUserIdsExists(userIds []int32) error {
 	}
 	return nil
 }
+
+func ExtractTokenNameFromEmail(email string) string {
+	return strings.Split(email, ":")[1]
+}
+
+func CheckIfTokenIsValid(email string, version string, userRepository repository.UserRepository) (bool, error) {
```
Comment (nishant-d) [inline pkg/auth/user/helper/helper.go:55]: we shouldn't make db call from helper/util

**Review by nishant-d — CHANGES_REQUESTED** [review]

**Review by nishant-d — DISMISSED** [review]

**Review by nishant-d — APPROVED** [review]

**Pull request merged**